model/fv_control.F90

```diff
@@ -452,6 +456,9 @@ subroutine fv_control_init(Atm, dt_atmos, this_grid, grids_on_this_pe, p_split,
      allocate(global_pelist(npes))
      call mpp_get_current_pelist(global_pelist, commID=global_commID) ! for commID
 
+     ! Need the fcst_mpi_comm in order to construct the communicators used by MPI_Bcast in fill_nested_grid_cpl()
+     call ESMF_VMGetCurrent(vm=vm,rc=rc)
```

Comment:
FMS does have a similar call (mpp_broadcast), but it creates the communicator on-the-fly every time its called.  My testing shows that the creation of the communicator takes about the same amount of time as the actual Bcast.  That's why I create the communicator in the fv_control_init().  Some numbers from an offline reproducer run on acorn.  The offline code creates the MPI communicator every iteration.  The first run takes a long time (cycles) probably due to MPI initialization.  "setup subcomm" refers to the call to MPI_Comm_split()

Cycles to setup subcomm: 845895398
Cycles Bcast with a comm: 352677465

Subsequent calls within the same execution settle into a nice pattern.
Cycles to setup subcomm: 35054707
Cycles Bcast with a comm: 3219772

Cycles to setup subcomm: 1966522
Cycles Bcast with a comm: 3727102

Cycles to setup subcomm: 2671380
Cycles Bcast with a comm: 3211785

Cycles to setup subcomm: 1973183
Cycles Bcast with a comm: 3195247

Cycles to setup subcomm: 1884240
Cycles Bcast with a comm: 3205283

Cycles to setup subcomm: 2135385
Cycles Bcast with a comm: 3193492

Cycles to setup subcomm: 2105010
Cycles Bcast with a comm: 3203685

Reply:
@dkokron - please send my your test program so I can confirm your findings with mpp_broadcast.  As I look at the code, if the pelist has been defined previously, it should be using a predefined communicator.  A new communicator is defined only when a new, unique pelist is encountered.

Reply:
It's on cactus.  I will upload when cactus comes back from dedicated time.

Perhaps then it's enough to create a pelist with the appropriate PEs in fv_control_init() instead of an MPI_Comm.

Reply:
Thanks for letting me know about the cactus downtime.

There should be no need to create the pelist in fv_control.  It can be defined/used within the fill_nested_grid_cpl function.  The call to get_peset [here](https://github.com/NOAA-GFDL/FMS/blob/main/mpp/include/mpp_transmit_mpi.fh#L177) will execute this code in [mpp_util_mpi.inc](https://github.com/NOAA-GFDL/FMS/blob/main/mpp/include/mpp_util_mpi.inc#L91-L139).  If the pelist is recognized, it will return the mpp-managed internal list location and mpp_broadcast will use the communicator associated with that list entry.  If the pelist isn't recognized, it will add it to the linked list of known pelists and create a communicator via MPI_Comm_create_group.  If you want to look over the code with me, let me know and we can schedule it.

Reply:
I see your point about being able to use get_peset() to create the communicator once storing it in the pset structure, then mpp_broadcast() reuses that communicator.  My implementation does essentially the same thing while storing the communicator in the fv_atmos_type structure.

I'm a little worried about the extra work done in the MPP approach.  Every call to get_peset() will check for monotonicity and 
check for a match to an existing peset.  I'll see if I can code up the MPP approach in my unit tester and determine if my worry is justified.

Reply:
My worry was about the extra work was unfounded.  I will re-code the PR using mpp_broadcast.

You can find my unit testers under clogin02:/u/daniel.kokron/Bcast/
The testers simulate what happens in a HAFS case with one nest run on 736 ranks.

Some timings:
grep Cycles Bcast.o58131230

Using MPI_Bcast with a pre-made communicator (C implementation)
Cycles to setup subcomm: 789298313
 Cycles Bcast with a comm: 350650485
 Cycles Bcast with a comm: 96031373
 Cycles Bcast with a comm: 3138885
 Cycles Bcast with a comm: 3160485
 Cycles Bcast with a comm: 3150698
 Cycles Bcast with a comm: 3127387
 Cycles Bcast with a comm: 3128850
 Cycles Bcast with a comm: 3114293
 Cycles Bcast with a comm: 3123540
 Cycles Bcast with a comm: 3147165
 Cycles Bcast with a comm: 3136793
 Cycles Bcast with a comm: 3124845
 Cycles Bcast with a comm: 3137737
 Cycles Bcast with a comm: 3132450
 Cycles Bcast with a comm: 3156840

Using MPI_Bcast with a pre-made communicator (Fortran implementation)
 Cycles to setup subcomm    136254240
 Cycles rank/thread    332684775
 Cycles rank/thread      4624537
 Cycles rank/thread      3276248
 Cycles rank/thread      3257415
 Cycles rank/thread      3283357
 Cycles rank/thread      3271545
 Cycles rank/thread      3424860
 Cycles rank/thread      3280073
 Cycles rank/thread      3279398
 Cycles rank/thread      3288915
 Cycles rank/thread      3359723
 Cycles rank/thread      3296475
 Cycles rank/thread      3246457
 Cycles rank/thread      3312000
 Cycles rank/thread      3278633

Using mpp_broadcast() 
Cycles rank/thread    548630797
 Cycles rank/thread      3106642
 Cycles rank/thread      3273503
 Cycles rank/thread      3308108
 Cycles rank/thread      3264412
 Cycles rank/thread      3284077
 Cycles rank/thread      3473212
 Cycles rank/thread      3260768
 Cycles rank/thread      3294653
 Cycles rank/thread      3311325
 Cycles rank/thread      3297983
 Cycles rank/thread      3296722
 Cycles rank/thread      3306937
 Cycles rank/thread      3325117
 Cycles rank/thread      3270375

Reply:
@dkokron - I appreciate your willingness to work with us on this.

Reply:
Never easy.

Swapping in mpp_broadcast in place of MPI_Bcast in fill_nested_grid_cpl() results in runtime failure.

```
   subroutine fill_nested_grid_cpl(this_grid, proc_in)
-    use mpi
+    !use mpi
+    use mpp_mod, only: mpp_broadcast
     integer, intent(in) :: this_grid
     logical, intent(in), optional :: proc_in
 
@@ -2460,7 +2461,8 @@ contains
                             g_dat(isg:,jsg:,1), position=CENTER)
     endif
     if(any(mpp_pe() == Atm(this_grid)%Bcast_ranks)) then
-      call MPI_Bcast(g_dat, size(g_dat), MPI_REAL, 0, Atm(this_grid)%Bcast_comm, ierr) ! root==0 because sending rank (Atm(this_grid)%sending) is rank zero in Bcast_comm
+      !call MPI_Bcast(g_dat, size(g_dat), MPI_REAL, 0, Atm(this_grid)%Bcast_comm, ierr) ! root==0 because sending rank (Atm(this_grid)%sending) is rank zero in Bcast_comm
+      call mpp_broadcast(g_dat, size(g_dat),Atm(this_grid)%sending_proc,Atm(this_grid)%Bcast_ranks)
     endif
     call timing_off('COMM_TOTAL')
     if (process) then
```
The stack trace points to a part of the code I didn't change.

> nid001001.acorn.wcoss2.ncep.noaa.gov 0: NOTE from PE     0:  eval_move_nest: move_cd_x=           0 move_cd_y=           0 do_move= F
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: NOTE from PE     0: Opened BC file: INPUT/gfs_bndy.tile7.003.nc
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: forrtl: error (78): process killed (SIGTERM)
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: Image              PC                Routine            Line        Source             
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000258B397  fv_regional_mod_m        3387  fv_regional_bc.F90
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000257BCD5  fv_regional_mod_m        1792  fv_regional_bc.F90
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000259E347  fv_regional_mod_m        1592  fv_regional_bc.F90
> nid001001.acorn.wcoss2.ncep.noaa.gov 0: fv3_32bit.exe      000000000291AABA  atmosphere_mod_mp         673  atmosphere.F90